### PR TITLE
[k8s] Show hints when requested resources don't fit in Kubernetes cluster

### LINKED ISSUE #3506: [k8s] [GKE] Fail to request T4 instance
I'm running a single T4 node on GKE. Nodes are properly labeled as shown below and `sky show-gpus --cloud kubernetes` is also correct but fails to launch.

```bash 
(sky) gcpuser@gfd-ebd1-head-evggxnxq-compute:~/skypilot$ kubectl describe nodes | grep cloud.google.com/gke-accelerator
                    cloud.google.com/gke-accelerator=nvidia-tesla-t4
                      cloud.google.com/gke-accelerator=nvidia-tesla-t4,cloud.google.com/gke-boot-disk=pd-balanced,cloud.google.com/gke-container-runtime=contain...
(sky) gcpuser@gfd-ebd1-head-evggxnxq-compute:~/skypilot$ sky show-gpus --cloud kubernetes
COMMON_GPU  AVAILABLE_QUANTITIES  
T4          1                     

Hint: use -a/--all to see all accelerators (including non-common ones) and pricing.
(sky) gcpuser@gfd-ebd1-head-evggxnxq-compute:~/skypilot$ sky launch --cloud kubernetes --gpus T4
I 05-02 08:48:17 optimizer.py:1209] No resource satisfying Kubernetes({'T4': 1}) on Kubernetes.
sky.exceptions.ResourcesUnavailableError: Kubernetes cluster does not contain any instances satisfying the request:
Task<name=sky-cmd>(run=<empty>)
  resources: Kubernetes({'T4': 1}).

To fix: relax or change the resource requirements.

Hint: sky show-gpus to list available accelerators.
      sky check to check the enabled clouds.
```

_Version & Commit info:_
* `sky -v`: `skypilot, version 1.0.0-dev0`
* `sky -c`: `skypilot, commit 889adce65602b76e31f60534ce25c264bad7cb83`

### PR DESCRIPTION
**Pull request by romilbhardwaj**
This PR adds the plumbing to surface hints from the cloud when no feasible resources are found. This is useful for surfacing errors in Kubernetes.

Closes #3506.

### Examples
```
===== GPU drivers not installed on the node =====
sky launch -c test3 --gpus T4:1 --cloud kubernetes
I 05-23 13:39:13 optimizer.py:1263] No resource satisfying Kubernetes({'T4': 1}) on Kubernetes.
I 05-23 13:39:13 optimizer.py:1272] Kubernetes: Could not detect GPU resources (`nvidia.com/gpu`) in Kubernetes cluster. If this cluster contains GPUs, please ensure GPU drivers are installed on the node. Check if the GPUs are setup correctly by running `kubectl describe nodes` and looking for the nvidia.com/gpu resource. Please refer to the documentation on how to set up GPUs.
sky.exceptions.ResourcesUnavailableError: Kubernetes cluster does not contain any instances satisfying the request:
Task<name=sky-cmd>(run=<empty>)
  resources: Kubernetes({'T4': 1}).

To fix: relax or change the resource requirements.

Hint: sky show-gpus to list available accelerators.
      sky check to check the enabled clouds.


======= When enough CPU is not available =======
$ sky launch -c test3 --gpus T4:1 --cpus 32 --cloud kubernetes
I 05-23 13:47:27 optimizer.py:1263] No resource satisfying Kubernetes(cpus=32, {'T4': 1}) on Kubernetes.
I 05-23 13:47:27 optimizer.py:1272] Kubernetes: GPU nodes with T4 do not have enough CPU (> 32.0 vCPUs) and/or memory (> 128.0 G). Maximum resources found on a single node: 16.0 CPUs, 58.9G Memory
sky.exceptions.ResourcesUnavailableError: Kubernetes cluster does not contain any instances satisfying the request:
Task<name=sky-cmd>(run=<empty>)
  resources: Kubernetes(cpus=32, {'T4': 1}).

To fix: relax or change the resource requirements.

Hint: sky show-gpus to list available accelerators.
      sky check to check the enabled clouds.
```

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Manual tests from above issue and other optimizer tests in tests/*

### CODE REVIEW

File: sky/clouds/aws.py [hunk 1]
```diff
@@ -425,7 +425,7 @@ def make_deploy_resources_variables(self,
 
     def _get_feasible_launchable_resources(
         self, resources: 'resources_lib.Resources'
-    ) -> Tuple[List['resources_lib.Resources'], List[str]]:
+    ) -> Tuple[List['resources_lib.Resources'], List[str], Optional[str]]:
```
Comment (Michaelvll) [inline sky/clouds/aws.py:428]: Seems the return value is getting longer. Could we also do a refactoring for it in this PR, e.g., creating a data class for the return type to make it simpler?
```
@dataclasses.dataclass
class FeasibleResourcesWithHints:
    pass
```

**Review by Michaelvll — COMMENTED** [review]
Thanks for adding this @romilbhardwaj! This is great, as for some clouds, if we specify a specific `cpus: 10` without `+` at the end, many clouds will not be able to provide such instance, having a hint for that would be very useful. Can we add a TODO for adding those hints in the future?

File: sky/clouds/aws.py (hunk 1 above)
Comment (romilbhardwaj) [reply sky/clouds/aws.py]: Should be fixed now! Added `resource_utils.FeasibleResources`.

**Comment by romilbhardwaj** [comment]
Thanks @Michaelvll - I've refactored the feasible resources returned by clouds into a single dataclass. Ready for another look.

File: sky/optimizer.py [hunk 2]
```diff
@@ -1239,21 +1235,25 @@ def _fill_in_launchable_resources(
...
+        # If clouds provide hints, store them for later printing.
+        hints: Dict[clouds.Cloud, str] = {}
         for cloud in clouds_list:
-            (feasible_resources,
-             fuzzy_candidate_list) = cloud.get_feasible_launchable_resources(
-                 resources, num_nodes=task.num_nodes)
-            if len(feasible_resources) > 0:
+            feasible_resources = cloud.get_feasible_launchable_resources(
+                resources, num_nodes=task.num_nodes)
+            if feasible_resources.hint:
```
Comment (Michaelvll) [inline sky/optimizer.py:1243]:
```suggestion
            if feasible_resources.hint is not None:
```

File: sky/clouds/aws.py [hunk 3]
```diff
@@ -452,10 +452,10 @@ def _get_feasible_launchable_resources(
                 region=resources.region,
                 zone=resources.zone)
             if not regions:
-                return ([], [])
+                return resources_utils.FeasibleResources([], [], None)
```
Comment (Michaelvll) [inline sky/clouds/aws.py:455]: I am leaning towards adding a `TODO` in these clouds that returns `None`, to add more hints for why a resource has failed to be found. (Our failover reasons sometimes are not helpful enough, and these additional hints should significantly help be set properly).

**Review by Michaelvll — APPROVED** [review]
Thanks for fixing this @romilbhardwaj! It looks good to me to go in if the backward compatibility tests pass. : )

File: sky/clouds/aws.py (hunk 3 above)
Comment (romilbhardwaj) [reply sky/clouds/aws.py:457]: Agreed, added a TODO comment for each cloud.

**Comment by romilbhardwaj** [comment]
Tested backward compatibility with manual tests, merging now.

**Pull request merged**